pkg/manager/member/pvc_cleaner.go

```diff
@@ -40,37 +47,144 @@ type PVCCleanerInterface interface {
 	pvcControl controller.PVCControlInterface
 	pvcLister  corelisters.PersistentVolumeClaimLister
+	pvLister   corelisters.PersistentVolumeLister
+	pvControl  controller.PVControlInterface
 }
 
 // NewRealPVCCleaner returns a realPVCCleaner
 func NewRealPVCCleaner(
+	kubeCli kubernetes.Interface,
 	podLister corelisters.PodLister,
 	pvcControl controller.PVCControlInterface,
-	pvcLister corelisters.PersistentVolumeClaimLister) PVCCleanerInterface {
+	pvcLister corelisters.PersistentVolumeClaimLister,
+	pvLister corelisters.PersistentVolumeLister,
+	pvControl controller.PVControlInterface) PVCCleanerInterface {
 	return &realPVCCleaner{
+		kubeCli,
 		podLister,
 		pvcControl,
 		pvcLister,
+		pvLister,
+		pvControl,
 	}
 }
 
 func (rpc *realPVCCleaner) Clean(tc *v1alpha1.TidbCluster) (map[string]string, error) {
+	if skipReason, err := rpc.cleanScheduleLock(tc); err != nil {
+		return skipReason, err
+	}
+
+	if tc.Spec.DeferPVCDelete {
+		// defer PVC delete policy is enabled, can't reclaim PV, so return directly.
+		return nil, nil
+	}
+
+	return rpc.reclaimPV(tc)
+}
+
+func (rpc *realPVCCleaner) reclaimPV(tc *v1alpha1.TidbCluster) (map[string]string, error) {
 	ns := tc.GetNamespace()
 	tcName := tc.GetName()
+
 	// for unit test
 	skipReason := map[string]string{}
 
-	selector, err := label.New().Instance(tc.GetLabels()[label.InstanceLabelKey]).Selector()
+	pvcs, err := rpc.listAllPVCs(tc)
 	if err != nil {
-		return skipReason, fmt.Errorf("cluster %s/%s assemble label selector failed, err: %v", ns, tcName, err)
+		return skipReason, err
 	}
 
-	pvcs, err := rpc.pvcLister.PersistentVolumeClaims(ns).List(selector)
+	for _, pvc := range pvcs {
+		pvcName := pvc.GetName()
+		l := label.Label(pvc.Labels)
+		if !(l.IsPD() || l.IsTiKV()) {
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotPDOrTiKV
+			continue
+		}
+
+		if pvc.Status.Phase != corev1.ClaimBound {
+			// If pvc is not bound yet, it will not be processed
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotBound
+			continue
+		}
+
+		if len(pvc.Annotations[label.AnnPVCDeferDeleting]) == 0 {
+			// This pvc has not been marked as defer delete PVC, can't reclaim the PV bound to this PVC
+			skipReason[pvcName] = skipReasonPVCCleanerIsNotDeferDeletePVC
+			continue
+		}
+
+		// PVC has been marked as defer delete PVC, try to reclaim the PV bound to this PVC
+		podName, exist := pvc.Annotations[label.AnnPodNameKey]
+		if !exist {
+			// PVC has not pod name annotation, this is an unexpected PVC, skip it
+			skipReason[pvcName] = skipReasonPVCCleanerPVCNotHasPodNameAnn
+			continue
+		}
+
+		_, err := rpc.podLister.Pods(ns).Get(podName)
+		if err == nil {
+			// PVC is still referenced by this pod, can't reclaim PV
+			skipReason[pvcName] = skipReasonPVCCleanerPVCeferencedByPod
+			continue
+		}
+		if !errors.IsNotFound(err) {
+			return skipReason, fmt.Errorf("cluster %s/%s get pvc %s pod %s from local cache failed, err: %v", ns, tcName, pvcName, podName, err)
```

Comment:
Why `return` for this one error for one PVC instead of `continue`? Same for the following returns.

Reply:
The logic here is that if this error occurs, then we have reason to believe that the next operation is likely to fail, so we let it fail quickly and wait for the next sync cycle.